### DISCUSSION TITLE
Extract frames from GIF, APNG, and animated WEBP

### DISCUSSION THREAD

**Discussion by Jay123210599** [ImageMagick/ImageMagick#7158] [Help]
### ImageMagick version

7.1.1-27

### Operating system, version and so on

Windows 11

### Description

How do I extract frames from GIF, APNG, and animated WEBP files?

**Comment by fmw42** [comment]
image.suffix[0] is the first frame image.suffix[1] is the second frame, etc

**Comment by snibgo** [comment]
As fmw42 says. For example:
```
magick in.gif[0-3] out.png
```
... will read the first 4 frames from a GIF files, and write them to out-0.png, out-1.png, etc.

**Comment by dceeee** [comment]
using ImageMagick 7.1.1-20 Q16-HDRI x64 98bb1d4:20231008 on Windows11
I find that these statements 
magick in.gif out.png
magick in.webp out.png
magick apng:in.png out.png
will extract all frames

I cannot (yet) get the [N] or [N-n] to extract specific frames

**Reply by fmw42** [reply]
What was your command to extract a given frame?  Does the input have more than one frame?  What does magick identify image.suffix show?

**Comment by snibgo** [comment]
> I cannot (yet) get the [N] or [N-n] to extract specific frames

Show your input file, and your command, and any output text messages the command generated.

**Comment by dceeee** [comment]
Extraordinary, I tried again this morning and the [] worked for gif & webp files, even magick moulin.gif[0,2,5-8,13] out.png
Png's are another matter, I think my moulin.png is a bit mucky so I started again using
Wiki_beach_ball.png from https://commons.wikimedia.org/wiki/File:Animated_PNG_example_bouncing_beach_ball.png

magick apng:Wiki_beach_ball.png -format "%N" info:   20
magick apng:Wiki_beach_ball.png[0-3] out.png  4 files
magick apng:Wiki_beach_ball.png[2-6] out.png  first frame only
magick apng:Wiki_beach_ball.png[6-8] out.png  error magick: invalid image index `Wiki_beach_ball.png' @ error/list.c/CloneImages/296
magick apng:Wiki_beach_ball.png[0-8] out.png  9 files
magick apng:Wiki_beach_ball.png[1-8] out.png  6 files
magick apng:Wiki_beach_ball.png[2-8] out.png  3 files

following all produce the same first frame
magick apng:Wiki_beach_ball.png[0] out0.png
magick apng:Wiki_beach_ball.png[6] out6.png
magick apng:Wiki_beach_ball.png[12] out12.png

I'd guess that this is because magick is using ffmpeg and the syntax is different, I'm rather rusty with ffmpeg so will need to refresh
The Wiki file is public domain so it's included in the zip for convenience together with mine
[Wiki_beach_ballAndMoulin.zip](https://github.com/ImageMagick/ImageMagick/files/14558721/Wiki_beach_ballAndMoulin.zip)

**Comment by snibgo** [comment]
> following all produce the same first frame

I confirm the problem.

We can see what images are in the zipped file (using IM 7.1.1-28):
```
magick apng:Wiki_beach_ball.png info:

apng:Wiki_beach_ball.png[0] APNG 100x100 100x100+0+0 8-bit sRGB 53682B 0.000u 0:00.000
apng:Wiki_beach_ball.png[1] APNG 39x63 100x100+30+36 8-bit sRGB 0.003u 0:00.002
apng:Wiki_beach_ball.png[2] APNG 39x56 100x100+30+40 8-bit sRGB 0.005u 0:00.004
apng:Wiki_beach_ball.png[3] APNG 39x65 100x100+30+34 8-bit sRGB 0.006u 0:00.005
apng:Wiki_beach_ball.png[4] APNG 39x77 100x100+30+22 8-bit sRGB 0.007u 0:00.007
apng:Wiki_beach_ball.png[5] APNG 39x81 100x100+30+18 8-bit sRGB 0.008u 0:00.008
apng:Wiki_beach_ball.png[6] APNG 39x83 100x100+30+16 8-bit sRGB 0.009u 0:00.009
apng:Wiki_beach_ball.png[7] APNG 39x85 100x100+30+14 8-bit sRGB 0.011u 0:00.010
apng:Wiki_beach_ball.png[8] APNG 39x91 100x100+30+8 8-bit sRGB 0.012u 0:00.011
apng:Wiki_beach_ball.png[9] APNG 39x97 100x100+30+2 8-bit sRGB 0.013u 0:00.012
apng:Wiki_beach_ball.png[10] APNG 39x37 100x100+30+2 8-bit sRGB 0.014u 0:00.013
apng:Wiki_beach_ball.png[11] APNG 39x36 100x100+30+2 8-bit sRGB 0.015u 0:00.014
apng:Wiki_beach_ball.png[12] APNG 39x37 100x100+30+2 8-bit sRGB 0.016u 0:00.016
apng:Wiki_beach_ball.png[13] APNG 39x97 100x100+30+2 8-bit sRGB 0.018u 0:00.017
apng:Wiki_beach_ball.png[14] APNG 39x95 100x100+30+4 8-bit sRGB 0.019u 0:00.018
apng:Wiki_beach_ball.png[15] APNG 39x91 100x100+30+8 8-bit sRGB 0.020u 0:00.020
apng:Wiki_beach_ball.png[16] APNG 39x89 100x100+30+10 8-bit sRGB 0.021u 0:00.021
apng:Wiki_beach_ball.png[17] APNG 39x85 100x100+30+14 8-bit sRGB 0.022u 0:00.022
apng:Wiki_beach_ball.png[18] APNG 39x79 100x100+30+20 8-bit sRGB 0.024u 0:00.023
apng:Wiki_beach_ball.png[19] APNG 39x75 100x100+30+24 8-bit sRGB 0.025u 0:00.024
```
So [6] should be 39x83 pixels:
```
magick apng:Wiki_beach_ball.png[6] info:

apng:Wiki_beach_ball.png[6] APNG 100x100 100x100+0+0 8-bit sRGB 3780B 0.000u 0:00.000
```

`-verbose` tells us the ffmpeg command:
```
magick -verbose apng:Wiki_beach_ball.png[6] +verbose info:

"ffmpeg.exe" -nostdin -loglevel error -i "C:/Users/Alan/AppData/Local/Temp/magick-Ut-gKHHQn-dSlOaPwd0CmaAx5An9DmZg" -an -f rawvideo -y -vframes 1 -pix_fmt rgba -vcodec webp -lossless 1 "C:/Users/Alan/AppData/Local/Temp/magick-Rxr4Wrxcr9nk6KqqlT-BoS8AW1BB5CjR"
```
The ffmpeg command seems to be wrong.

**Comment by dceeee** [comment]
As
magick apng:Wiki_beach_ball.png[0,2,5-8,13] out.png
doesn't currently work, the ffmpeg equivalent is

ffmpeg -i Wiki_beach_ball.png -vf select='eq(n,0)+eq(n,2)+between(n,5,8)+eq(n,13)' -vsync 0 out%02d.png

This works for windows, other platforms may need the commas to be escaped eg eq(n\,0)  -edit: there should be a slash between the n & comma  but it doesn't show up

**Comment by Earnestly** [comment]
Is it possible to extract specific frames while also using `-define filename:literal=true`?

**Comment by snibgo** [comment]
See https://imagemagick.org/script/defines.php :
```
filename:literal=true 	By default, an output filename can contain embedded formatting characters.
Use this option to bypass interpreting embedded formatting characters and instead use the filename literally.
```
So this affects *output* filenames.

**Comment by Earnestly** [comment]
Oh in that case I have a few questions;

1. Is it possible to disable interpretation of input filenames? They are user supplied and I wish to avoid injection.
2. I initially thought it effected the input filename because `[0]` was showing up in `Thumb::URI`, but I have recently tried this again and noticed the issue seems unrelated:

```sh
$ magick -ping doc/art/tmux_status.gif\[0\] -thumbnail 1x1\< -format '%[thumb::uri]\n' info:
./doc%2Fart%2Ftmux_status.gif%5B0%5D
```

3. Given the above, should the URI encoding include `/` as `%2F` or should it be left as `/`? The examples in the specification don't appear to encode the <slash> (`/`):

> e.g. file:///home/jens/photo/me.jpg

4. Should I convert these into issues instead of posting here?

**Comment by snibgo** [comment]
> 1. Is it possible to disable interpretation of input filenames? They are user supplied and I wish to avoid injection.

What interpretation(s) do you want to disable? The `[]` suffix will match a file, if it exists. For example:
```
magick t.png[3] info:
```
This will succeed if IM finds a file named `t.png[3]`. If there is no such file, IM will look for a file named `t.png`.

> 3. Given the above, should the URI encoding include / as %2F or should it be left as /?

I'm not sure what you mean.

MagickCore/resize.c function ThumbnailImage() encodes the URL so converts each slash, then prepends either `file://` or `./`, so those slashes are not encoded. Is that what you mean?

> 4. Should I convert these into issues instead of posting here?

For questions about what IM does, this is the correct place. For requests to change what IM does, raise an issue. There is inevitably an overlap, eg "Why does IM do ABC, when I would expect XYZ?".

**Reply by Earnestly** [reply]
> > 1. Is it possible to disable interpretation of input filenames? They are user supplied and I wish to avoid injection.
> 
> What interpretation(s) do you want to disable? The `[]` suffix will match a file, if it exists. For example:
> 
> ```
> magick t.png[3] info:
> ```
> 
> This will succeed if IM finds a file named `t.png[3]`. If there is no such file, IM will look for a file named `t.png`.

Yes, I would want to disable all of this. If the filename was `t.png[3]` I would not want imagemagick to look for `t.png`, instead it should be a "file not found" error.

> > 3. Given the above, should the URI encoding include / as %2F or should it be left as /?
> 
> I'm not sure what you mean.
> 
> MagickCore/resize.c function ThumbnailImage() encodes the URL so converts each slash, then prepends either `file://` or `./`, so those slashes are not encoded. Is that what you mean?

I don't believe it should encode slashes. That is, `/foo/bar.png` should become `file:///foo/bar.png` and not `file://%2Ffoo%2Fbar.png`.

**Comment by snibgo** [comment]
The change so `t.png[3]` does not look for `t.png` could be implemented as a `-define`.

I suggest you raise two issues, as they are requests for changes, and they are not related. Try to explain *why* you want the changes, eg the problems created by the current behaviours.

**Comment by Earnestly** [comment]
The main issue is how to query a specific frame without using filename interpretation. Is this currently possible?

**Comment by snibgo** [comment]
As far as I know, the only way to specify specific input frames is with the `[...]` suffix to the filename.

**Comment by Earnestly** [comment]
Is it possible to select a specific frame while also using inline resize? E.g. `t.png[128x128][0]`? If not, is there a way to use inline resize without using `[...]` notation?

*(It is possible to use `-delete` for "selecting" a specific frame but is quite a bit slower.)*
